cpp/src/Slice/Parser.cpp

```diff
@@ -953,6 +953,35 @@ Slice::Contained::isTopLevel() const
     return dynamic_pointer_cast<Unit>(container()) != nullptr;
 }
 
+ModulePtr
+Slice::Contained::getTopLevelModule() const
+{
+    ContainedPtr p;
+
+    // Round-about way to get `ContainedPtr` for this element, since we can't `enable_shared_from_this`.
+    for (const auto& sibling : container()->contents())
+    {
+        if (sibling.get() == this)
+        {
+            p = sibling;
+        }
+    }
+    assert(p);
```

Comment:
Is there a better way to obtain a `ContainedPtr` of `this` here?

`Contained` doesn't `enable_shared_from_this` and cannot, so that it's subclasses can support `shared_from_this` with more specific typing.

Reply:
Did you consider defining this as a separate function `Slice::getTopLevelModule(ContainedPtr) const`?

Reply:
Yes, and I am still considering it! If no one has a better idea, I'll probably switch a free function.
But I figured I'd ask before falling back to it, since we tend to prefer member functions over free functions.

Reply:
I don't find this convoluted code acceptable.

If you want to use a member function, make it virtual and override it at least in Module. Then use recursion.

Reply:
I also don't find it acceptable, I was just showing what I wanted in hopes there was a 5* C++ way to accomplish it.
I'll simplify and switch to a free function, as suggested.

Reply:
Virtual member function and recursion seems pretty straightforward to me. Am I missing something?

Reply:
No, you're not.

Just that it would of been even more straightforward if I could get a `ContainedPtr` of this.
Then, I'd only need one implementation on `Contained`, instead of 2 (one for `Contained` and one for `Module`).

Either way, I changed the implementation as suggested.